ext/standard/mail.c

```diff
@@ -269,30 +268,21 @@ PHP_FUNCTION(mail)
 		Z_PARAM_STRING(subject, subject_len)
 		Z_PARAM_STRING(message, message_len)
 		Z_PARAM_OPTIONAL
-		Z_PARAM_ZVAL(headers)
+		Z_PARAM_STR_OR_ARRAY_HT(headers_str, headers_ht)
 		Z_PARAM_STR(extra_cmd)
 	ZEND_PARSE_PARAMETERS_END();
 
 	/* ASCIIZ check */
 	MAIL_ASCIIZ_CHECK(to, to_len);
 	MAIL_ASCIIZ_CHECK(subject, subject_len);
 	MAIL_ASCIIZ_CHECK(message, message_len);
-	if (headers) {
-		switch(Z_TYPE_P(headers)) {
-			case IS_STRING:
-				tmp_headers = zend_string_init(Z_STRVAL_P(headers), Z_STRLEN_P(headers), 0);
-				MAIL_ASCIIZ_CHECK(ZSTR_VAL(tmp_headers), ZSTR_LEN(tmp_headers));
-				str_headers = php_trim(tmp_headers, NULL, 0, 2);
-				zend_string_release_ex(tmp_headers, 0);
-				break;
-			case IS_ARRAY:
-				str_headers = php_mail_build_headers(headers);
-				break;
-			default:
-				php_error_docref(NULL, E_WARNING, "headers parameter must be string or array");
-				RETURN_FALSE;
-		}
+	if (headers_str) {
+		MAIL_ASCIIZ_CHECK(ZSTR_VAL(headers_str), ZSTR_LEN(headers_str));
+		headers_str = php_trim(headers_str, NULL, 0, 2);
+	} else if (headers_ht) {
```

Comment:
I would write this an else with `ZEND_ASSERT(headers_ht)` to make it clear that one of the two branches has to be hit.

Reply:
Hm, are you sure that one has to be hit? The `$headers` parameter is optional, so that's why I think the `else if` is needed.

Reply:
Ooops, I missed that it's optional :)